src/Microsoft.AspNet.Server.Kestrel/Http/SocketOutput.cs

```diff
@@ -138,6 +144,58 @@ public void End(ProduceEndType endType)
             }
         }
 
+        public MemoryPoolIterator2 ProducingStart()
+        {
+            lock (_returnLock)
+            {
+                Debug.Assert(!_isProducing);
+                _isProducing = true;
+
+                if (_tail == null)
+                {
+                    throw new IOException("The socket has been closed.");
+                }
+
+                return new MemoryPoolIterator2(_tail, _tail.End);
+            }
+        }
+
+        public void ProducingComplete(MemoryPoolIterator2 end, int count)
```

Comment:
Actually it might not literally require all of the blocks to be filled... As long as the block.End is updated in the iter.CopyFrom method, the SocketOutput can use the block .Begin and .End properties to know what regions are occupied by the data to be written